server/channels/app/authentication.go

```diff
@@ -147,31 +147,70 @@ func (a *App) DoubleCheckPassword(rctx request.CTX, user *model.User, password s
 	return nil
 }
 
-func (a *App) checkLdapUserPasswordAndAllCriteria(rctx request.CTX, ldapId *string, password string, mfaToken string) (*model.User, *model.AppError) {
-	if a.Ldap() == nil || ldapId == nil {
+func (a *App) checkLdapUserPasswordAndAllCriteria(rctx request.CTX, userID, loginID, password, mfaToken string) (*model.User, *model.AppError) {
+	// MM-37585: Use locks to avoid concurrently checking AND updating the failed login attempts.
+	a.ch.ldapLoginAttemptsMut.Lock()
+	defer a.ch.ldapLoginAttemptsMut.Unlock()
+
+	user, appErr := a.GetUserForLogin(rctx, userID, loginID)
+	if appErr != nil {
+		return nil, appErr
+	}
```

Comment:
That call adds an additional round trip to the LDAP server. Can we avoid that?

Reply:
Yeah I didn't have this originally, it's being used to ensure we have the correct value for `MaxAttempts` after we acquire the lock. I'll remove this call and just get the updated `MaxAttempts` value for a user